### ISSUE TITLE
Overflow Panic

### ISSUE THREAD

**Issue by rydrman** [DavidVentura/tpom#1]
Hello! I was giving this a try in a rust project but ran into a consistent panic when trying to execute the `ClockController::overwrite`. The line in question is below, I'm happy to help with a little debugging but my knowledge is pretty limited in this space so I'm hoping someone else might have an inclination on what's going wrong and/or what an appropriate solution would be.

https://github.com/DavidVentura/tpom/blob/370e3bade303f5cba308090f5c7a0f451fb2e121/src/vdso.rs#L210 

```
thread 'main' panicked at 'attempt to add with overflow', /home/rbottriell/.cargo/git/checkouts/tpom-0ec46c840464517e/370e3ba/src/vdso.rs:210:24
stack backtrace:
   0: rust_begin_unwind
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/panicking.rs:67:14
   2: core::panicking::panic
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/panicking.rs:117:5
   3: tpom::vdso::vDSO::overwrite
             at /home/rbottriell/.cargo/git/checkouts/tpom-0ec46c840464517e/370e3ba/src/vdso.rs:210:24
   4: tpom::ClockController::mess_vdso
             at /home/rbottriell/.cargo/git/checkouts/tpom-0ec46c840464517e/370e3ba/src/lib.rs:171:17
   5: tpom::ClockController::overwrite
             at /home/rbottriell/.cargo/git/checkouts/tpom-0ec46c840464517e/370e3ba/src/lib.rs:165:9
```

**Comment by DavidVentura** [comment]
How are you calling `overwrite`? Are you able to clone this repo and run `cargo run`?
What distro+kernel are you running? (you can check with `cat /etc/os-release` and `uname -r`)

**Comment by rydrman** [comment]
I can clone it, but get the same result:
```console
> cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.16s
     Running `target/debug/tpom`
Now: SystemTime { tv_sec: 1694637917, tv_nsec: 640102506 }
thread 'main' panicked at 'attempt to add with overflow', src/vdso.rs:210:24
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Here's my distro information:

```console
> cat /etc/centos-release
CentOS Linux release 7.9.2009 (Core)
> uname -r
3.10.0-1160.90.1.el7.x86_64
```

**Comment by DavidVentura** [comment]
Interesting, running with debug logs on my Debian machine I get 
```log
david@dev:~/tpom$ RUST_LOG=debug cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running `target/debug/tpom`
Now: SystemTime { tv_sec: 1694638799, tv_nsec: 666501105 }
[2023-09-13T20:59:59Z DEBUG tpom] Found vDSO starting at 0x7ffe9d30c000, ending at 0x7ffe9d30e000, for a size of 8192 (0x2000)
[2023-09-13T20:59:59Z DEBUG tpom::vdso] Overwriting elf_offset 0x7ffe9d30c000, symbol_address 0x960, with jmp_address 0x555f9445a1d0
[1970-01-01T00:00:01Z DEBUG tpom::vdso] Overwriting elf_offset 0x7ffe9d30c000, symbol_address 0x780, with jmp_address 0x555f9445a510
[1970-01-01T00:00:01Z DEBUG tpom::vdso] Overwriting elf_offset 0x7ffe9d30c000, symbol_address 0x780, with jmp_address 0x555f9445a510
[1970-01-01T00:00:01Z DEBUG tpom::vdso] Overwriting elf_offset 0x7ffe9d30c000, symbol_address 0x960, with jmp_address 0x555f9445a1d0
```

and on a CentOS7 VM

```log
Now: SystemTime { tv_sec: 1694638766, tv_nsec: 30975319 }
[2023-09-13T20:59:26Z DEBUG tpom] Found vDSO starting at 0x7ffdff88b000, ending at 0x7ffdff88d000, for a size of 8192 (0x2000)
[2023-09-13T20:59:26Z DEBUG tpom::vdso] Overwriting elf_offset 0x7ffdff88b000, symbol_address 0xffffffffff700600, with jmp_address 0x55da3b6f6c90
```

looks like the reporting for the memory maps is different!

**Comment by rydrman** [comment]
ah yes, that `symbol_address` definitely would be an issue - it's a bit beyond me, to be honest but do you have a sense of what the fix would be? Or a direction you can point me to look at if you'd like my help?

**Comment by DavidVentura** [comment]
I've just pushed a "fix" to a branch, in fbf85e52c0f6471eb2a06113cee91c747d5433e7, which you can apply locally to validate. It does work on my CentOS7 test machine:

```
$ uname -r
3.10.0-1160.95.1.el7.x86_64
$ cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running `target/debug/tpom`
Now: SystemTime { tv_sec: 1694641570, tv_nsec: 362692159 }
Executing
Now: SystemTime { tv_sec: 1, tv_nsec: 3 }
Now: SystemTime { tv_sec: 1694641570, tv_nsec: 363459017 }

```

I'll spend some time tomorrow looking at which field from the ELF symbol table I am not taking into proper consideration, instead of hardcoding an `& 0xffff`

**Comment by rydrman** [comment]
Amazing, thank you - I'll give it a go and see. So far no panics but it's not working as I expected (pretty sure this is my fault, I'll keep digging)

**Comment by DavidVentura** [comment]
This particular issue got fixed in 1a9c9b9b9fae83212a9ef7780482ce91e79488d5, thanks for reporting!
